sycl/test-e2e/Basic/large-range.cpp

```diff
@@ -0,0 +1,83 @@
+// RUN: %{build} -o %t.out
+// RUN: env SYCL_PARALLEL_FOR_RANGE_ROUNDING_TRACE=1 %{run} %t.out
+
+#include <numeric>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+
+constexpr specialization_id<int> C;
+int n_fail = 0;
+
+template <typename ContainerT>
+void check_sum(std::string_view desc, const ContainerT &data, size_t N) {
+  auto res = std::accumulate(data.begin(), data.end(), size_t(0));
+  if (res != N) {
+    std::cout << desc << " fail\n"
+              << "got:      " << res << "\n"
+              << "expected: " << N << "\n";
+    ++n_fail;
+  } else {
+    std::cout << desc << " pass\n";
+  }
+}
+
+template <typename RangeT>
+void test_regular(std::string_view desc, queue &q, size_t B, RangeT range) {
+  auto N = range.size();
+  std::vector accumulators_v(B, 0, usm_allocator<int, usm::alloc::shared>(q));
+  auto *accumulators = accumulators_v.data();
+
+  q.parallel_for(range, [=](auto it) {
+     atomic_ref<int, memory_order::relaxed, memory_scope::device> ref(
+         accumulators[it.get_linear_id() % B]);
+     ++ref;
+   }).wait();
+
+  check_sum(desc, accumulators_v, N);
+}
+
+template <typename RangeT>
+void test_spec_constant(std::string_view desc, queue &q, size_t B,
+                        RangeT range) {
+  auto N = range.size();
+  std::vector accumulators_v(B, 0, usm_allocator<int, usm::alloc::shared>(q));
+  auto *accumulators = accumulators_v.data();
+
+  q.submit([&](handler &cgh) {
+     cgh.set_specialization_constant<C>(2);
+     cgh.parallel_for(range, [=](auto it, kernel_handler h) {
+       atomic_ref<int, memory_order::relaxed, memory_scope::device> ref(
+           accumulators[it.get_linear_id() % B]);
+       ref += h.get_specialization_constant<C>();
+     });
+   }).wait();
+
+  check_sum(desc, accumulators_v, N * 2);
+}
+
+int main(int argc, char *argv[]) {
+  if (sizeof(size_t) <= 4) {
+    std::cout << "size_t is 32-bit, nothing to do\n";
+    return 0;
+  }
+
+  queue q;
+  int B = 1000000;
+  size_t N = 4'294'967'311;
```

Comment:
Can we use `numeric_limits`? And also use `constexpr` wherever possible.

Reply:
Not really, I just took this value from the CTS, and they say it's the first prime number bigger than UINT32_MAX. But I can add a comment above the line to justify the value.

Reply:
A comment would be nice, yes.